lang/types/value.go

```diff
@@ -153,6 +164,14 @@ func ValueOf(v reflect.Value) (Value, error) {
 		values := make(map[string]Value)
 		for i := 0; i < l; i++ {
 			x := value.Field(i)
+			// It's likely that setting the field is desirable, so clone the
+			// field to prevent the recursive ValueOf() call failing. This is
+			// obviously slower so hope we don't need this often.
+			if !x.CanSet() {
+				c := reflect.New(x.Type()).Elem()
```

Comment:
I'm still not 100% on this, but it seems to work as-is. I feel like I'm inferring `x.Kind() == Ptr` from `!x.CanSet()` which may or may not hold true in the majority of cases. I think the thing I'm not sure on is _why_ `x.CanSet() == false`. Debugging this isn't easy when it's all just pointers